lib/pos_fee_estimator.go

```diff
@@ -0,0 +1,506 @@
+	defer posFeeEstimator.rwLock.Unlock()
+
+	if err := posFeeEstimator.removeBlockNoLock(block); err != nil {
+		return errors.Wrap(err, "PoSFeeEstimator.RemoveBlock: error removing block from PoSFeeEstimator")
+	}
+	return nil
+}
+
+// removeBlockNoLock is the same as RemoveBlock but assumes the caller has already acquired the rwLock.
+func (posFeeEstimator *PoSFeeEstimator) removeBlockNoLock(block *MsgDeSoBlock) error {
+	// Remove all transaction from the block from the pastBlocksTransactionRegister.
+	for _, txn := range block.Txns {
+		// We explicitly exclude block reward transactions as they do not have fees.
+		// They were never added in the first place.
+		if txn.TxnMeta.GetTxnType() == TxnTypeBlockReward {
+			continue
+		}
+		mtxn, err := NewMempoolTx(txn, block.Header.TstampNanoSecs, block.Header.Height)
+		if err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.RemoveBlock: error creating MempoolTx")
+		}
+		if err = posFeeEstimator.pastBlocksTransactionRegister.RemoveTransaction(mtxn); err != nil {
+			return errors.Wrap(err,
+				"PoSFeeEstimator.removeBlockNoLock: error removing txn from pastBlocksTransactionRegister")
+		}
+	}
+	blockHash, err := block.Hash()
+	if err != nil {
+		return errors.Wrap(err, "PoSFeeEstimator.removeBlockNoLock: error computing blockHash")
+	}
+	// Remove the block from the cached blocks.
+	for ii, cachedBlock := range posFeeEstimator.cachedBlocks {
+		cachedBlockHash, err := cachedBlock.Hash()
+		if err != nil {
+			return errors.Wrap(err, "PoSFeeEstimator.removeBlockNoLock: error computing cachedBlockHash")
+		}
+		if blockHash.IsEqual(cachedBlockHash) {
+			posFeeEstimator.cachedBlocks = append(posFeeEstimator.cachedBlocks[:ii],
+				posFeeEstimator.cachedBlocks[ii+1:]...)
+			break
+		}
+	}
+	return nil
+}
+
+// UpdateGlobalParams updates the global params used by the PoSFeeEstimator. This only modifies the GlobalParams
+// used by the pastBlockTransactionRegister and allows it to properly compute the fee buckets. The mempool's
+// global params are not modified and are controlled externally.
+func (posFeeEstimator *PoSFeeEstimator) UpdateGlobalParams(globalParams *GlobalParamsEntry) error {
+	posFeeEstimator.rwLock.Lock()
+	defer posFeeEstimator.rwLock.Unlock()
+	tempTransactionRegister := NewTransactionRegister(globalParams.Copy())
+	for _, block := range posFeeEstimator.cachedBlocks {
+		if err := addBlockToTransactionRegister(tempTransactionRegister, block); err != nil {
+			return errors.Wrap(err, "PosFeeEstimator.UpdateGlobalParams: error adding block to tempTransactionRegister")
+		}
+	}
+	return nil
+}
+
+// sortCachedBlocks sorts the cached blocks by height & tstamp just to be safe.
+func (posFeeEstimator *PoSFeeEstimator) sortCachedBlocks() {
+	posFeeEstimator.cachedBlocks = collections.SortStable(posFeeEstimator.cachedBlocks,
+		func(ii, jj *MsgDeSoBlock) bool {
+			if ii.Header.Height == jj.Header.Height {
+				return ii.Header.TstampNanoSecs < jj.Header.TstampNanoSecs
```

Comment:
Maybe add one more tie-breaker on block hash just so that the sort is 100% stable.